src/lib/reasoners/rel_utils.ml

```diff
@@ -193,126 +193,179 @@ end = struct
-                  assert false
-              ) parents
-          ) t.parents (X.leaves r)
+  let pp ppf { cs_set; cs_map = _ ; fresh = _ } =
+    Fmt.(
+      braces @@ hvbox @@
+      iter ~sep:semi CS.iter @@
+      box ~indent:2 @@ braces @@
+      Constraint.pp
+    ) ppf cs_set
+
+  let empty =
+    { cs_set = CS.empty
+    ; cs_map = MX.empty
+    ; fresh = CS.empty }
+
+  let cs_add cs r cs_map =
+    MX.update r (function
+        | Some css -> Some (CS.add cs css)
+        | None -> Some (CS.singleton cs)
+      ) cs_map
+
+  let cs_remove cs r cs_map =
+    MX.update r (function
+        | Some css ->
+          let css = CS.remove cs css in
+          if CS.is_empty css then None else Some css
+        | None ->
+          (* Can happen if the same argument is repeated *)
+          None
+      ) cs_map
+
+  let subst ex rr nrr bcs =
+    match MX.find rr bcs.cs_map with
+    | ids ->
+      let cs_map, cs_set, fresh =
+        CS.fold (fun cs (cs_map, cs_set, fresh) ->
+            let fresh = CS.remove cs fresh in
+            let cs_set = CS.remove cs cs_set in
+            let cs_map = Constraint.fold_leaves (cs_remove cs) cs cs_map in
+            let cs' = Constraint.subst ex rr nrr cs in
+            if CS.mem cs' cs_set then
+              cs_map, cs_set, fresh
+            else
+              let cs_set = CS.add cs' cs_set in
+              let cs_map = Constraint.fold_leaves (cs_add cs') cs' cs_map in
+              (cs_map, cs_set, CS.add cs' fresh)
+          ) ids (bcs.cs_map, bcs.cs_set, bcs.fresh)
       in
-      { parents ; registered = SX.remove r t.registered }
-    else
-      invalid_arg "Congruence.remove"
-
-  let subst rr nrr cgr f t =
-    match MX.find rr cgr.parents with
-    | rr_deps ->
-      let cgr = { cgr with parents = MX.remove rr cgr.parents } in
-      SX.fold (fun r (cgr, t) ->
-          let r' = X.subst rr nrr r in
-          (* [r] contains [rr] as a leaf by definition *)
-          assert (not (X.equal r r'));
-
-          (* Update the other leaves *)
-          let parents =
-            List.fold_left (fun parents other_leaf ->
-                if X.equal other_leaf rr then
-                  parents
-                else
-                  MX.update other_leaf (function
-                      | Some deps ->
-                        let deps = SX.remove r deps in
-                        if SX.is_empty deps then None else Some deps
-                      | None -> assert false
-                    ) parents
-              ) cgr.parents (X.leaves r)
-          in
-
-          (* Add the new representative to the congruence if needed *)
-          let cgr = add r' { cgr with parents } in
+      assert (not (MX.mem rr cs_map));
+      { cs_set ; cs_map ; fresh }
+    | exception Not_found -> bcs
 
-          (* Propagate the substitution *)
-          cgr, f r r' t
-        ) rr_deps (cgr, t)
-    | exception Not_found -> cgr, t
+  let add bcs c =
+    if CS.mem c bcs.cs_set then
```

Comment:
We use the polymorphic comparison here. It means the explanation is updated by this function but the `Constraint.compare` is supposed to ignore the explanation? Is it intentional?

We should at least mention in the documentation of `add` that the structure is updated if the explanation changed.

Reply:
Where do we use polymorphic comparison? `CS` is defined as `Set.Make(Constraint)`, `CS.mem` uses `Constraint.compare`.